pkg/healthcheck/healthcheck.go

```diff
@@ -200,28 +200,47 @@ func (hc *HealthChecker) addKubernetesAPIChecks() {
 }
 
 func (hc *HealthChecker) addLinkerdPreInstallChecks() {
-	hc.checkers = append(hc.checkers, &checker{
-		category:    LinkerdPreInstallCategory,
-		description: "control plane namespace does not already exist",
-		check: func() error {
-			exists, err := hc.kubeAPI.NamespaceExists(hc.httpClient, hc.ControlPlaneNamespace)
-			if err != nil {
-				return err
-			}
-			if exists {
-				return fmt.Errorf("The \"%s\" namespace already exists", hc.ControlPlaneNamespace)
-			}
-			return nil
-		},
-	})
+	if hc.SingleNamespace {
+		hc.checkers = append(hc.checkers, &checker{
+			category:    LinkerdPreInstallCategory,
+			description: "control plane namespace exists",
+			check: func() error {
+				exists, err := hc.kubeAPI.NamespaceExists(hc.httpClient, hc.ControlPlaneNamespace)
+				if err != nil {
+					return err
+				}
+				if !exists {
+					return fmt.Errorf("The \"%s\" namespace must exist before installing with --single-namespace", hc.ControlPlaneNamespace)
+				}
+				return nil
+			},
+		})
+	} else {
+		hc.checkers = append(hc.checkers, &checker{
+			category:    LinkerdPreInstallCategory,
+			description: "control plane namespace does not already exist",
+			check: func() error {
+				exists, err := hc.kubeAPI.NamespaceExists(hc.httpClient, hc.ControlPlaneNamespace)
+				if err != nil {
+					return err
+				}
+				if exists {
+					return fmt.Errorf("The \"%s\" namespace already exists", hc.ControlPlaneNamespace)
+				}
+				return nil
+			},
+		})
+	}
 
-	hc.checkers = append(hc.checkers, &checker{
-		category:    LinkerdPreInstallCategory,
-		description: "can create Namespaces",
-		check: func() error {
-			return hc.checkCanCreate("", "", "v1", "Namespace")
-		},
-	})
+	if !hc.SingleNamespace {
+		hc.checkers = append(hc.checkers, &checker{
```

Comment:
tioli you could put this in the if/else right above, to avoid splitting out the logic for "if single namespace" vs "if not single namespace"?

Reply:
In addition to @rmars's comment, do you think it is possible to DRY this check up a little bit? Something like this?
```diff
diff --git a/pkg/healthcheck/healthcheck.go b/pkg/healthcheck/healthcheck.go
index 49a479c9..082f5f8f 100644
--- a/pkg/healthcheck/healthcheck.go
+++ b/pkg/healthcheck/healthcheck.go
@@ -200,35 +200,30 @@ func (hc *HealthChecker) addKubernetesAPIChecks() {
 }
 
 func (hc *HealthChecker) addLinkerdPreInstallChecks() {
+	namespaceCheck := func(msg string) func() error {
+		return func() error {
+		exists, err := hc.kubeAPI.NamespaceExists(hc.httpClient, hc.ControlPlaneNamespace)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf(msg, hc.ControlPlaneNamespace)
+		}
+		return nil
+		}
+	}
+
 	if hc.SingleNamespace {
 		hc.checkers = append(hc.checkers, &checker{
 			category:    LinkerdPreInstallCategory,
 			description: "control plane namespace exists",
-			check: func() error {
-				exists, err := hc.kubeAPI.NamespaceExists(hc.httpClient, hc.ControlPlaneNamespace)
-				if err != nil {
-					return err
-				}
-				if !exists {
-					return fmt.Errorf("The \"%s\" namespace must exist before installing with --single-namespace", hc.ControlPlaneNamespace)
-				}
-				return nil
-			},
+			check:       namespaceCheck("The \"%s\" namespace must exist before installing with --single-namespace"),
 		})
 	} else {
 		hc.checkers = append(hc.checkers, &checker{
 			category:    LinkerdPreInstallCategory,
 			description: "control plane namespace does not already exist",
-			check: func() error {
-				exists, err := hc.kubeAPI.NamespaceExists(hc.httpClient, hc.ControlPlaneNamespace)
-				if err != nil {
-					return err
-				}
-				if exists {
-					return fmt.Errorf("The \"%s\" namespace already exists", hc.ControlPlaneNamespace)
-				}
-				return nil
-			},
+			check:       namespaceCheck("The \"%s\" namespace already exists"),
 		})
 	}
```

Reply:
All good suggestions -- have addressed them in the latest commit. Turns out we already had a `checkNamespace` func in this file that I could repurpose for these checks.